### PR TITLE
Allow guild commands in seperate guilds to have duplicate names

### PR DESCRIPTION
**Pull request by LordOfPolls**
## About this pull request

In the current form of the library, you cannot create a guild command if another guild/global has a command with the same name, this pr resolves this, and allows for duplicate names. 

This change **will** break any command that uses `self.commands` or `self.subcommands` (ie help commands) but the fix is simple. Otherwise, every effort has been made to prevent breaking changes 

## Changes

`slash.commands` and `slash.subcommands` were modified to store a dict of each guild (incl global) and their respective commands
rather than storing commands like so:
```json
"commandName": {"command Data"}
```
its stored like this:
```json
guildID:"commandName":{"command Data"}
```
The functions that add, process, and call commands have all been modified to respect tis change, and through my testing work correctly

Some variables have also been renamed from abstract names (like `x` `y` `z`) to more descriptive names, if they were related to my patch

## Checklist

- [x] I've checked this pull request runs on `Python 3.6.X`.
- [ ] This fixes something in [Issues](https://github.com/eunwoo1104/discord-py-slash-command/issues).
    - Issue:
- [x] This adds something new.
- [x] There is/are breaking change(s).
- [ ] (If required) Relavent documentation has been updated/added.
- [ ] This is not a code change. (README, docs, etc.)

### CODE REVIEW

File: discord_slash/client.py [hunk 1]
```diff
@@ -417,12 +415,19 @@ def add_subcommand(self,
...
         description = description or getdoc(cmd)
-        guild_ids = guild_ids if guild_ids else []
-
-        if base in self.commands:
-            for x in guild_ids:
-                if x not in self.commands[base].allowed_guild_ids:
-                    self.commands[base].allowed_guild_ids.append(x)
+        guild_ids = guild_ids if guild_ids else ["global"]
+
+        for i in range(len(guild_ids)):
```
Comment (quirky-bluejay) [inline discord_slash/client.py:432]:
```suggestion
        for guild in guild_ids:
```

File: discord_slash/client.py [hunk 2]
```diff
@@ -417,12 +415,19 @@ def add_subcommand(self,
...
+            if guild_ids[i] is None:
+                guild_ids[i] = "global"
+            if guild_ids[i] not in self.commands:
+                self.commands[guild_ids[i]] = {}
+            if guild_ids[i] not in self.subcommands:
+                self.subcommands[guild_ids[i]] = {}
+            if base in self.commands[guild_ids[i]]:
+                for x in guild_ids:
+                    if x not in self.commands[guild_ids[i]][base].allowed_guild_ids:
+                        self.commands[guild_ids[i]][base].allowed_guild_ids.append(x)
```
Comment (quirky-bluejay) [inline discord_slash/client.py:442]: Pretty sure this could cause issues
Because you could have for example:
```
/base
 - cmd1 - guild1
 - cmd2 - guild2
 - cmd3 - guild1 + guild2
 ```
 Not actually sure that it would but it seems like it might? (this is confusing)

File: discord_slash/client.py [hunk 3]
```diff
@@ -445,27 +450,28 @@ def add_subcommand(self,
...
-            obj = model.SubcommandObject(_sub, base, name, subcommand_group)
-            self.subcommands[base][subcommand_group][name] = obj
-        else:
-            if name in self.subcommands[base]:
-                raise error.DuplicateCommand(f"{base} {name}")
-            obj = model.SubcommandObject(_sub, base, name)
-            self.subcommands[base][name] = obj
-        self.logger.debug(f"Added subcommand `{base} {subcommand_group or ''} {name or cmd.__name__}`")
+        for guild in guild_ids:
+            if base not in self.commands:
```
Comment (quirky-bluejay) [inline discord_slash/client.py]: did you mean?
```suggestion
            if base not in self.commands[guild]:
```

File: discord_slash/client.py [hunk 4]
```diff
@@ -600,7 +606,8 @@ async def _group_kick_user(ctx, user):
         subcommand_group_description = subcommand_group_description or sub_group_desc
 
         def wrapper(cmd):
-            obj = self.add_subcommand(cmd, base, subcommand_group, name, description, base_description, subcommand_group_description, guild_ids, options, connector)
+            obj = self.add_subcommand(cmd, base, subcommand_group, name, description, base_description,
+                                      subcommand_group_description, guild_ids, options, connector)
```
Comment (quirky-bluejay) [inline discord_slash/client.py:622]: Why has this changed other than formatting (which probably should happen, but be in a different pr), there are no other changes here :upside_down_face:

File: discord_slash/client.py [hunk 5]
```diff
@@ -717,39 +724,49 @@ async def on_socket_response(self, msg):
...
+        ctx = context.SlashContext(self.req, to_use, self._discord, self.logger)
+        cmd_name = to_use["data"]["name"]
 
-            if cmd_name not in self.commands and cmd_name in self.subcommands:
+        # check guild commands and global commands for this command
+        if ctx.guild_id in self.commands and \
+                to_use["data"]["name"] in self.commands[ctx.guild_id]:
+            selected_cmd = self.commands[ctx.guild_id][to_use["data"]["name"]]
+        elif to_use["data"]["name"] in self.commands["global"]:
+            selected_cmd = self.commands["global"][to_use["data"]["name"]]
```
Comment (quirky-bluejay) [inline discord_slash/client.py]: pretty sure that this'll select a guild command to run if it exists, even if a global command was called, with the same name, because `ctx.guild_id` is only none in dms
I've had a look at the [docs](https://discord.com/developers/docs/interactions/slash-commands#interaction) regarding the response received from discord, and it looks like they don't tell you if it's global or guild though

File: discord_slash/client.py [hunk 6]
```diff
@@ -761,9 +778,15 @@ async def handle_subcommand(self, ctx: context.SlashContext, data: dict):
         :param ctx: :class:`.model.SlashContext` instance.
         :param data: Gateway message.
         """
-        if data["data"]["name"] not in self.subcommands:
+        if ctx.guild_id in self.subcommands:
+            if data["data"]["name"] in self.subcommands[ctx.guild_id]:
+                local = ctx.guild_id
+        elif data["data"]["name"] in self.subcommands["global"]:
+            local = "global"
```
Comment (quirky-bluejay) [inline discord_slash/client.py]: Same with comment i said above in `on_socket_response`

**Review by quirky-bluejay — CHANGES_REQUESTED** [review]
I've highlighted a few things I think need tweaking.
Unfortunately I think I have come across an issue with figuring out whether what was received was global or guild specific, when two commands exist in both with the same names (comment on `on_socket_response`

File: discord_slash/client.py [hunk 7]
```diff
@@ -328,10 +320,6 @@ def add_slash_command(self,
         """
         Registers slash command to SlashCommand.
 
-        .. warning::
-            Just using this won't register slash command to Discord API.
-            To register it, check :meth:`.utils.manage_commands.add_slash_command` or simply enable `sync_commands`.
-
```
[the remark this answers is not in the document]
Comment (LordOfPolls) [reply discord_slash/client.py]: Honestly dont even remember removing it, probably got accidentally removed at some point, ill add it back

File: discord_slash/client.py [hunk 8]
```diff
@@ -349,15 +337,23 @@ def add_slash_command(self,
...
-            tgt = self.commands[name]
-            if not tgt.has_subcommands:
-                raise error.DuplicateCommand(name)
-            has_subcommands = tgt.has_subcommands
-            for x in tgt.allowed_guild_ids:
-                if x not in guild_ids:
-                    guild_ids.append(x)
+        guild_ids = guild_ids if guild_ids else ["global"]
+
+        for i in range(len(guild_ids)):
```
[the remark this answers is not in the document]
Comment (LordOfPolls) [reply discord_slash/client.py]: I chose to use `i in range` as I'm updating the list of which I'm iterating over

File: discord_slash/client.py (hunk 4 above)
Comment (LordOfPolls) [reply discord_slash/client.py:622]: `Ctrl Alt + L` in pycharm reformats your code, I use it a lot when coding, I can revert it, but technically it abides by PEP8 now

File: discord_slash/client.py (hunk 1 above)
Comment (LordOfPolls) [reply discord_slash/client.py:432]: same as before
I chose to use `i in range` as I'm updating the list of which I'm iterating over

File: discord_slash/client.py [hunk 9]
```diff
@@ -206,89 +214,69 @@ async def to_dict(self):
...
-            else:
-                if "global" not in wait:
-                    wait["global"] = {}
-                wait["global"][x] = copy.deepcopy(command_dict)
+        for guild in self.commands:
+            for command in self.commands[guild].values():
+                command_dict = {
+                    "name": command.name,
+                    "description": command.description or "No Description",
+                    "options": command.options or []
```
[the remark this answers is not in the document]
Comment (LordOfPolls) [reply discord_slash/client.py]: yep

File: discord_slash/client.py [hunk 10]
```diff
@@ -206,89 +214,69 @@ async def to_dict(self):
...
-        for x in self.commands:
-            for i in self.commands[x].allowed_guild_ids:
-                if i not in all_guild_ids:
-                    all_guild_ids.append(i)
+
+        output = {}
         cmds = {
             "global": [],
-            "guild": {x: [] for x in all_guild_ids}
+            "guild": {x: [] for x in self.commands.keys() if x != "global"}
```
[the remark this answers is not in the document]
Comment (LordOfPolls) [reply discord_slash/client.py:228]: I chose not to alter this output as it minimised changes in the rest of the library

File: discord_slash/client.py [hunk 11]
```diff
@@ -164,30 +170,32 @@ def remove_cog_commands(self, cog):
...
-                del self.commands[x.name]
+        for cmd_obj in res:
+            if isinstance(cmd_obj, model.CogCommandObject):
+                for guild in self.commands:
+                    if cmd_obj.name not in self.commands[guild]:
+                        continue  # Just in case it is removed due to subcommand.
+                    if cmd_obj.name in self.subcommands[guild]:
+                        self.commands[guild][cmd_obj.name].func = None
+                        continue  # Let's remove completely when every subcommand is removed.
+                    del self.commands[guild][cmd_obj.name]
```
[the remark this answers is not in the document]
Comment (LordOfPolls) [reply discord_slash/client.py]: Fixed in incoming commit

File: discord_slash/client.py [hunk 12]
```diff
@@ -349,15 +337,23 @@ def add_slash_command(self,
...
+                self.commands[guild_ids[i]] = {}
+            if name in self.commands[guild_ids[i]]:
+                # prevent duplicate commands in guild
+                tgt = self.commands[guild_ids[i]][name]
+                if not tgt.has_subcommands:
+                    raise error.DuplicateCommand(name)
+                has_subcommands = tgt.has_subcommands
+                for x in tgt.allowed_guild_ids:
+                    if x not in guild_ids:
+                        guild_ids.append(x)
```
[the remark this answers is not in the document]
Comment (LordOfPolls) [reply discord_slash/client.py:368]: In this case, I'm mirroring the original code, I simply kept the original code as it seemed to be its intent

**Comment by LordOfPolls** [comment]
Global commands now override guild commands where names are duplicated

File: discord_slash/client.py (hunk 4 above)
Comment (i0bs) [reply discord_slash/client.py:622]: PEP8 isn't really a concern to me IMO, code readability > formatting.

File: discord_slash/client.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (quirky-bluejay) [reply discord_slash/client.py]: Pretty certain that that won't work, because the length of the list is got once at the start of the loop, and therefore if you append anything to it the for loop won't get to there, because it's still only iterating the length of the original list.
Tested this with this:
```py
x = [1,2,3,4,5]
for i in range(len(x)):
    print(x[i])
    x.append(i+5)
print(x)
``` 
And yeah it doesn't work

File: discord_slash/client.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (quirky-bluejay) [reply discord_slash/client.py:368]: What about removing the `guild_ids` attribute?

File: discord_slash/client.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (LordOfPolls) [reply discord_slash/client.py]: It is at this moment I realised that I'm following rules for another language. I forgot python can handle modifying the list you're iterating over XD

File: discord_slash/client.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (LordOfPolls) [reply discord_slash/client.py:368]: aye, certainly could remove it, but im trying to minimise the scope of the changes i make. Right now, a user could update their code to use this lib without issues (excld `slash.commads`). Removing `guild_ids` would be more breaking which im trying to avoid

**Comment by LordOfPolls** [comment]
This is back in sync with master, and changes have been incorporated. Request testing by others
